File: Data Structures & Algorithms/max-area-of-island/submission-1.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0])
        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        max_area = 0
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 1:
                    q = deque([(r, c)])
                    grid[r][c] = 0
                    area = 0
                    while q:
                        r, c = q.popleft()
                        area += 1
                        for dr, dc in directions:
                            nr, nc = r + dr, c + dc
                            if 0<=nr<rows and 0<=nc<cols and grid[nr][nc] == 1:
                                grid[nr][nc] = 0
                                q.append((nr, nc))
                    max_area = max(max_area, area)
        return max_area


        

File: Data Structures & Algorithms/count-number-of-islands/submission-1.java
class Solution {

    public boolean isValid(int rows, int cols, int r, int c){
        return r >= 0 && r < rows && c >= 0 && c < cols; 
    }

    public void dfs(char[][] grid, int r, int c){
        int rows = grid.length;
        int cols = grid[0].length;

        int[][] directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

        for(int i = 0; i < 4; i++){
            int dr = directions[i][0];
            int dc = directions[i][1];

            int nr = r + dr;
            int nc = c + dc;
            if(isValid(rows, cols, nr, nc) && grid[nr][nc] == '1'){
                grid[nr][nc] = '0';
                dfs(grid, nr, nc);
            }
        }
    }
    public int numIslands(char[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;

        int numIslands = 0;
        for(int r = 0; r < rows; r++){
            for(int c = 0; c < cols; c++){
                if(grid[r][c] == '1'){
                    grid[r][c] = '0';
                    dfs(grid, r, c);
                    numIslands++;
                }
            }
        }

        return numIslands;
        
    }
}
